operator-wg/whitepaper-community-review/README.md

```diff
@@ -0,0 +1,880 @@
+
+Main advantages of metacontroller :
+* Only a function (called via webhook) without any boilerplate related to watching kubernetes resources needs to be provided
+* Such a function can be written in any language, and exposed via http
+
+Main limitations :
+* Only certain patterns are possible to implement, mentioned above
+* The current architecture relies on a single metacontroller in a cluster
+* Metacontroller is not aware of any external state, it relies entirely on cluster state
+
+Example metacontroller configuration, shown below, is used to add additional network exposure for `StatefulSet` without explicitly defining `Service` manifest. 
+```yaml
+apiVersion: metacontroller.k8s.io/v1alpha1
+kind: DecoratorController
+metadata:
+  name: service-per-pod
+spec:
+  resources:
+  - apiVersion: apps/v1
+    resource: statefulsets
+    annotationSelector:
+      matchExpressions:
+      - {key: service, operator: Exists}
+      - {key: port, operator: Exists}
+  attachments:
+  - apiVersion: v1
+    resource: services
+  hooks:
+    sync:
+      webhook:
+        url: http://service-per-pod.metacontroller/sync-service-per-pod
+        timeout: 10s
+
+```
+With above configuration :
+* `metacontroller`, for every object matching `spec.resources` description (in this case - `apps/v1/statefulsets` with `service` and `port` annotations), will watch for any change in matching objects (create/update/delete) and invoke `hooks.sync` on each of those
+* the `hooks.sync` can return objects which are described in `spec.attachement` (in this case - `v1/services`) which will be created/updated/deleted by `metacontroller`, according to `hook` response
+For example, if below `Statefulset` will be deployed:
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  annotations:
+    service: "statefulset.kubernetes.io/pod-name"
+    ports: "80:8080"
+...
+```
+given `Service` object will be created by metacontroller:
+```yaml
+apiVersion: "v1"
+kind: "Service"
+spec:
+  selector: "statefulset.kubernetes.io/pod-name"
+  ports:
+  - port: 80
+    targetPort: 8080
+```
+
+The user defined endpoint (in this example - `http://service-per-pod.metacontroller/sync-service-per-pod`) only needs to care about the calculation of the `Service` and how it should look like for a given `StatefulSet`.
+
+Additional examples and ideas what could be implemented using metacontroller, can be found at the [metacontroller-examples](https://metacontroller.github.io/metacontroller/examples.html) page !
+
+For any question, please visit our slack channel ([#metacontroller](https://kubernetes.slack.com/archives/CA0SUPUDP)) or ask it on [github discussions](https://github.com/metacontroller/metacontroller/discussions/).
+
```

Comment:
As a Python alternative to the normal Golang-based operator frameworks is given, it might be nice to show some love to https://github.com/flant/shell-operator - it makes operator building much more accessible to more Devops/Sysadmin type folks rather than developers.

Reply:
Feel free to add it, if you like ... I think this would add some value to the paper as it illustrates that operators could also be written in bash :-)